keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,285 @@
+---
+title: Certificates API
+authors:
+  - "@mikednaese"
+owning-sig: sig-auth
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-06-07
+last-updated: 2019-06-07
+status: implementable
+---
+
+# Certificates API
+
+## Table of Contents
+
+- [Certificates API](#certificates-api)
```

Comment:
Can you also add a section on alternatives, e.g. ACME?

Reply:
ACME didn't really exist when we started working on this so it wasn't considered. https://github.com/kubernetes/kubernetes/issues/2303#issuecomment-62776961 and https://github.com/kubernetes/kubernetes/issues/3168 predate the first draft of ACME by about a year. https://github.com/kubernetes/kubernetes/pull/20439 (when the only client we were thinking of was Kubelet) was merged after Draft 01. The final Draft 18 of ACME along with many revisions wasn't published until December 2018. There are various incompatibilities with the final ACME protocol that make it difficult or impossible to implement the feature set we need fully compliantly but it probably [doesn't matter](https://mailarchive.ietf.org/arch/msg/acme/XsPJjRxZLc7NVGVfnjHd2HHcJt8) if there exists users that would benefit. It is just a wire protocol, so it's possible to implement a frontend compatible with ACME even if not fully compliant. There are likely some ACME clients on github that could use it. It's hard to say without user journeys but e.g. a volume type that hides the API interaction altogether might be more broadly useful.

Reply:
Thanks for the history on this.  It is really helpful to understand how we got here :)